### PR TITLE
ENH: Add Kendall's tau-a and tau-c variants to scipy.stats.kendalltau()

### PR DESCRIPTION
**Pull request by vkk800**
This PR adds support for Kendall's tau-a and tau-b to the existing `scipy.stats.kendalltau()` function. Some unit tests are also added.

Two points that whoever looks into merging this should probably know:

* As far as I understood the theory behind the coefficients, the p-value should always be the same for all three variants since it is computed using only the difference between concordant and discordant pairs. Please correct me if someone has better knowledge on this.

* It seems that there is a separate implementation of `kendalltau()` for masked arrays in `mstats_basic.py`. This PR only changes the function that is used for non-masked arrays. It seems that even in the existing implementation there was no parity between the two. This PR, however, makes the tau-a variant of the masked array implementation accessible from `scipy.stats.kendalltau()` (which is wasn't before).

EDIT: This PR references this issue I submitted a few days ago: https://github.com/scipy/scipy/issues/9329

### CODE REVIEW

File: scipy/stats/stats.py
Comment (chrisb83) [inline scipy/stats/stats.py]: Maybe call the parameter type instead of variant, with possible values a, b, c

File: scipy/stats/stats.py
Comment (chrisb83) [inline scipy/stats/stats.py]: Inconsistent use of Tau and tau

File: scipy/stats/stats.py
Comment (chrisb83) [inline scipy/stats/stats.py]: Mention default. 

"all all" -> all

File: scipy/stats/stats.py
Comment (chrisb83) [inline scipy/stats/stats.py]: Suggest tau_ for readability

File: scipy/stats/stats.py
Comment (chrisb83) [inline scipy/stats/stats.py]: Else? I.e. what if variant == 'xyz'? Suggest to mention that nans are in the data in the error message

**Review by chrisb83 — COMMENTED** [review]
Thanks for the PR. I added a few comments. Tests look quite comprehensive. 
We need to sure about p-values for a and c.

One general remark: i guess one could try to avoid code duplication in stats and mstats by moving the main calculation to _stats_mstats_common.py

File: scipy/stats/stats.py
Comment (ilayn) [reply scipy/stats/stats.py]: It's better to write the type as `str, optional` and then give the default choice and options.

File: scipy/stats/stats.py
Comment (vkk800) [reply scipy/stats/stats.py]: I don't like the idea of overriding the built-in `type()` function of Python like this.

File: scipy/stats/stats.py
Comment (vkk800) [reply scipy/stats/stats.py]: The way I've written it seems to be consistent with other parts of scipy. I will, however, add the default choice.

**Comment by vkk800** [comment]
I read Kendall (1970) derivation of the significance tests for tau-a/b/c with some care. It says after introducing the formula that my current PR uses for all three p-values: "These equations provide us with tests of significance of the appropriate t_a, t_b or t_c coefficients, or rather of the values of S from which they are derived." However, it is clear from the derivation that this p-value is in fact the p-value conditional to there being a given number of ties in the set.

One could also imagine deriving a probability distribution for the number of ties and incorporating this into the p-value calculation. However, if we are just given a dataset, this is impossible: for being able to tell what is the probability for having a certain number of ties in the set, we have to know how many different values the random variable where the sample is drawn from can get (if it can get only one value, then obviously all samples are always ties. If it can get infinite number of different values, then the probability of getting any ties at all is zero).

In short: the formula currently in the code gives one kind of p-value for the statistic, but one could also imagine computing a different kind of p-value that doesn't treat the number of ties as given. However, this is, in general, impossible if we don't have some additional information about the nature of the random variable being sampled.

**Comment by vkk800** [comment]
I pushed a new commit fixing some of the problems above. I did not change the p-value computation since I'm inclined to trust the way Kendall (1970) treats it, even though, as I stated above, it is not probably the only way to look at the issue.

**Comment by mlynarik** [comment]
Hi, do you know any estimate when this feature might be released (pull request merged)? Thank you.

**Comment by rgommers** [comment]
Thanks for the ping @mlynarik.  We lost track of it, the review queue is long (apologies @vkk800). I marked it for the next release so we'll make sure to get back to it before then.

**Comment by vkk800** [comment]
@mlynarik I think that probably the p-value computation needs to be reviewed with some care before this should be merged. I spent approximately one afternoon looking into this when I submitted my previous comment and I couldn't easily find any literature formula for p-values of tau-a/b/c separately. Using the same p-value for all of them comes with the caveats I described above. I don't really know what is the correct way to address the problem; it would be better if someone with a bit more background in rank correlation methods could give some input.

**Comment by mlynarik** [comment]
In my opinion, the p-value is not critical to be included if it’s not fully validated. I think that the sole Kendall Tau-A formula would be of great benefit for the package especially for those of us trying to calculate Somers’D.

**Comment by chrisb83** [comment]
Wikipedia indicates that tau_a and tau_b have different asymptotic distributions (without giving a reference):

https://en.wikipedia.org/wiki/Kendall_rank_correlation_coefficient#Significance_tests

The SAS doc also states different variances for tau-c and tau-b to be used when computed the ratio est / std(est).

So maybe one could figure out the right formulas from theses documents but looks like quite a bit of work. exact p-val calculation does not work if there are ties I guess.

Maybe returning nan for tau-a and -c is acceptable and the p-values can be added in a follow-up PR.

**Comment by tylerjereddy** [comment]
@chrisb83 @WarrenWeckesser @rlucas7 Pinging some "stats floks"--if we want this in `1.5.0` it needs to be merge ready in ~1 week, but has been inactive for ~6 months.

File: scipy/stats/stats.py
Comment (mdhaber) [reply scipy/stats/stats.py]: What do you think of :math:`tau` throughout?

File: scipy/stats/stats.py
Comment (mdhaber) [reply scipy/stats/stats.py:3464]: I think I agree with "variant" or, as I was thinking, "version". But maybe variant is best.

File: scipy/stats/stats.py
Comment (mdhaber) [reply scipy/stats/stats.py]: I agree that the set notation is used throughout scipy.stats.

File: scipy/stats/stats.py
Comment (mdhaber) [reply scipy/stats/stats.py]: And how about rendering this, too?

File: scipy/stats/stats.py
Comment (mdhaber) [reply scipy/stats/stats.py]: `else` was added. 
Re: nans, that is handled by `_contains_nan`

**Comment by mdhaber** [comment]
> The SAS doc also states different variances for tau-c and tau-b to be used when computed the ratio est / std(est).

@chrisb83 The SAS manual references Brown and Benedetti, which states:
image
and then even more explicitly about tau-b and tau-c:
image

Also, @vkk800 noted
> However, if we are just given a dataset, this is impossible: for being able to tell what is the probability for having a certain number of ties in the set, we have to know how many different values the random variable where the sample is drawn from can get

I had been having similar thoughts for Page's L test, but I agree that it seems standard to ignore that. 

So I think that the *p*-value we have here is correct - or standard, anyway - for tau-c and tau-b.

File: scipy/stats/tests/test_stats.py
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:837]: Why specify `variant='taub'`? Should it be raised for all variants?

**Comment by mdhaber** [comment]
@chrisb83 
> Wikipedia indicates that tau_a and tau_b have different asymptotic distributions (without giving a reference):

I added Wikipedia's formula for the tau-a *p*-value:
```
    n = size
    za = 3*con_minus_dis / np.sqrt(n*(n-1)*(2*n+5)/2)
    p = 2*distributions.norm.sf(np.abs(za))
```
and in the absence of ties, it gives the same *p*-value as the KendallTau. As we might expect, it gives a different value in the presence of ties.

So I suppose the question is whether it makes sense to allow the user to choose 'tau-a' manually. If we do, then the user is able to get the tau-a statistic and p-value even when there are ties, but what is the point? It is not appropriate to use tau-a when there are ties; that's why tau-b was developed.

I suggest that we just remove the explicit `"taua"` variant entirely. If the user selects tau-b, they will get the tau-b statistic and p-value. In the absence of ties, these are equivalent to the tau-a statistic and p-value. If the user selects tau-c, they will get the tau-c statistic, but we can still use the existing code to calculate the p-value because Brown and Benedetti state that the hypothesis tests for tau-b and tau-c are identical.

What do you think of that and the suggestions above for rendering the math symbols? Let me know what you think and I can make the changes.

**Comment by swallan** [comment]
We can compare the outputs of this function to other tools. TLDR: This PR implementation matches tau in R and SAS for these examples, with some variations in p value. 

## Test 1
```
In [69]: x = [5, 2, 1, 3, 6, 4, 7]
    ...: y = [5, 2, 6, 3, 1, 7, 4]
    ...: stats.kendalltau(x,y)
Out[69]: KendalltauResult(correlation=-0.14285714285714288, pvalue=0.7726190476190476)
``` 
### SAS:
tau: -0.14286
p-value: 0.6523

### R:
tau: -0.1428571
p-value: .6523

## Test 2
```
In [70]: x = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
In [71]: y = [0, 2, 1, 3, 4, 6, 5, 7, 8, 9]
In [72]: stats.kendalltau(x,y)
Out[72]: KendalltauResult(correlation=0.911111111111111, pvalue=2.9761904761904762e-05)
``` 
### SAS:
tau: .91111
p-value: .0002

### R:
tau: 0.9111111
p-value: 0.0002453


## Test 3
```
x = [1, 1, 2]

y = [1, 2, 3]

stats.kendalltau(x,y)
Out[75]: KendalltauResult(correlation=0.816496580927726, pvalue=0.22067136191984682)
``` 
### SAS:
tau: 0.8165
p-value: 0.2207

### R:
tau: 0.8164966
p-value: 0.2207



## R code: 

```

cor.test(as.numeric(list(5,   2, 1, 3, 6, 4, 7)), as.numeric(list(5, 2, 6, 3, 1, 7, 4)),   method="kendall", exact=1)
--



cor.test(as.numeric(list(0,   1, 2, 3, 4, 5, 6, 7, 8, 9)), as.numeric(list(0, 2, 1, 3, 4, 6, 5, 7, 8, 9)),   method="kendall", exact=1)
--



cor.test(as.numeric(list(1,   1, 2)), as.numeric(list(1,2,3)), method="kendall", exact=1)
--



```

## SAS code
```

data   datatable;
--
input x y;
datalines;
5 5
2 2
1 6
3 3
6 1
4 7
7 4
;
title 'sas kendall';
proc print data=datatable;
proc corr data=datatable   KENDALL;

-- -


data   datatable;
--
input x y;
datalines;
0 0
1 2
2 1
3 3
4 4
5 6
6 5
7 7
8 8
9 9
;
 
proc corr KENDALL;


data   datatable;
--
input x y;
datalines;
1 1
2 1
2 3
;
 
proc corr KENDALL;



--- 



```

**Comment by mdhaber** [comment]
Thanks @swallan. Please try to use SAS `PROC FREQ` to get the tau-c p-value. I think that would settle the debate about whether the p-values are the same for tau-b and tau-c.

**Comment by swallan** [comment]
@mdhaber I ended up figuring it out. Here is the document that helped me, for future use: http://www.math.wpi.edu/saspdf/stat/chap28.pdf. I'll update the comment to indicate the new usage. 

Specifically about seeing tau-b and tau-c: 

```
data example1;
input x y;
cards;
1 1 
1 2 
2 3
;

proc print data=example1;
proc freq data = example1;
tables y*x;
test kentb STUTC ;
```
Screen Shot 2020-07-21 at 12 24 43 AM

So, the `table` specification is regarding the output data, not necessarily input.

**Comment by mdhaber** [comment]
Ah, ok, thank you!
It is good to see that the tau-b and tau-c p-values are identical, and these numbers for ASE and confidence intervals will help me check some code I was working on. Thanks!

**Comment by WarrenWeckesser** [comment]
I think @swallan and @mdhaber have already done similar comparisons, but just in case there is any question about the discrepancy in the p-values shown above, I repeated @swallan's "Test 1" and made sure that the methods used in `kendalltau` and R's `cor.test` were the same.  The "Test 1" data are

    x = [5, 2, 1, 3, 6, 4, 7]
    y = [5, 2, 6, 3, 1, 7, 4]

Here's `kendalltau`, using the default variant (i.e. tau-b):

```
In [127]: x = [5, 2, 1, 3, 6, 4, 7]

In [128]: y = [5, 2, 6, 3, 1, 7, 4]

In [129]: kendalltau(x, y, method='exact')
Out[129]: KendalltauResult(correlation=-0.14285714285714288, pvalue=0.7726190476190476)

In [130]: kendalltau(x, y, method='asymptotic')
Out[130]: KendalltauResult(correlation=-0.14285714285714288, pvalue=0.6523041372117631)
```
Here's the result of `cor.test` in R.  To the precision shown in the R output, the statistic and p-value match the result of `kendalltau`:

```
> x = c(5, 2, 1, 3, 6, 4, 7)
> y = c(5, 2, 6, 3, 1, 7, 4)
> cor.test(x, y, method='kendall', exact=TRUE)

    Kendall's rank correlation tau

data:  x and y
T = 9, p-value = 0.7726
alternative hypothesis: true tau is not equal to 0
sample estimates:
       tau 
-0.1428571 

> cor.test(x, y, method='kendall', exact=FALSE)

    Kendall's rank correlation tau

data:  x and y
z = -0.45056, p-value = 0.6523
alternative hypothesis: true tau is not equal to 0
sample estimates:
       tau 
-0.1428571 

```

For tau-c, the R package DescTools has the function `StuartTauC`.  It returns the statistic and optionally a confidence interval, but it doesn't return a p-value.  Here's a quick check of the statistic.  (`StuartTauC` accepts a contingency table, so we'll have to unpack the table into individual samples to do the corresponding calculation with `kendalltau`.)

```
> tab <- as.table(rbind(c(26,26,23,18,9), c(6,7,9,14,23)))
> tab
   A  B  C  D  E
A 26 26 23 18  9
B  6  7  9 14 23
> StuartTauC(tab)
[1] 0.4110953
```

Here's the Python version, with a helper function that converts the contingency table to an array of individual samples.

```
In [131]: def table_to_sample(c):
     ...:     i, j = np.indices(c.shape)
     ...:     pairs = list(zip(i.ravel(), j.ravel()))
     ...:     data = np.repeat(pairs, c.ravel(), axis=0)
     ...:     return data
     ...:   

In [132]: tab = np.array([[26, 26, 23, 18, 9], [6, 7, 9, 14, 23]])

In [133]: tab
Out[133]: 
array([[26, 26, 23, 18,  9],
       [ 6,  7,  9, 14, 23]])

In [134]: data = table_to_sample(tab)
In [135]: data.shape
Out[135]: (161, 2)

In [136]: kendalltau(data[:, 0], data[:, 1], variant='c')
Out[136]: KendalltauResult(correlation=0.4110952509548243, pvalue=1.846976725439985e-06)
```
The statistics computed by `StuartTauC` and `kendalltau` with `variant='c'` match.

File: scipy/stats/stats.py
Comment (WarrenWeckesser) [inline scipy/stats/stats.py:4343]: Typo:
```suggestion
    in how they are normalized to lie within the range -1 to 1; the
```

File: scipy/stats/stats.py
Comment (WarrenWeckesser) [inline scipy/stats/stats.py:4402]: Nitpick: use backticks around variables consistently.

File: scipy/stats/stats.py
Comment (WarrenWeckesser) [inline scipy/stats/stats.py:4500]: PEP 8: line too long.  A copy-edited version of the line:
```suggestion
         raise ValueError(f"Unknown variant of the method chosen: {variant}. "
                          "variant must be 'b' or 'c'.")
```

File: scipy/stats/stats.py
Comment (WarrenWeckesser) [inline scipy/stats/stats.py:4505]: PEP 8: line too long.
```suggestion
    # The p-value calculation is the same for all variants since the p-value
    # depends only on con_minus_dis.
```

File: scipy/stats/tests/test_stats.py
Comment (WarrenWeckesser) [inline scipy/stats/tests/test_stats.py:1073]: PEP 8: lines too long.
```suggestion
        assert_equal(stats.kendalltau([2, 2, 2], [2, 2, 2], variant=taux),
                     (np.nan, np.nan))
        assert_equal(stats.kendalltau([2, 0, 2], [2, 2, 2], variant=taux),
                     (np.nan, np.nan))
        assert_equal(stats.kendalltau([2, 2, 2], [2, 0, 2], variant=taux),
                     (np.nan, np.nan))
```

File: scipy/stats/tests/test_stats.py
Comment (WarrenWeckesser) [inline scipy/stats/tests/test_stats.py:1092]: PEP 8: lines too long.
```suggestion
    assert_approx_equal(stats.kendalltau([1, 1, 2], [1, 1, 2], variant='b')[0],
                        1.0)
    assert_approx_equal(stats.kendalltau([1, 1, 2], [1, 1, 2], variant='c')[0],
                        0.88888888)
```

**Review by WarrenWeckesser — CHANGES_REQUESTED** [review]
Looks good.  I requested a few style/PEP 8 fixes in-line, but otherwise I think this is ready to go.

File: scipy/stats/stats.py
Comment (WarrenWeckesser) [inline scipy/stats/stats.py:4402]:
```suggestion
    number of unique values in either `x` or `y`, whichever is smaller.
```

**Review by WarrenWeckesser — DISMISSED** [review]

File: scipy/stats/stats.py
Comment (WarrenWeckesser) [inline scipy/stats/stats.py:4501]: Fix indentation.
```suggestion
        raise ValueError(f"Unknown variant of the method chosen: {variant}. "
                         "variant must be 'b' or 'c'.")
```

File: scipy/stats/stats.py
Comment (WarrenWeckesser) [inline scipy/stats/stats.py:4335]:
```suggestion
def kendalltau(x, y, initial_lexsort=None, nan_policy='propagate',
               method='auto', variant='b'):
```

**Comment by mdhaber** [comment]
I'll merge when CI finishes.

**Comment by WarrenWeckesser** [comment]
@mdhaber, I was making a few more tweaks via the web interface, but I found a few more PEP-8 issues.   Making more tweaks via the web interface is a nuisance, so I'll make the rest of my changes locally, and push them one new commit.

**Review by WarrenWeckesser — APPROVED** [review]
All my changes were minor copy-edits or PEP-8 fixes.  The original PR with @mdhaber's updates look good, so +1.

**Comment by WarrenWeckesser** [comment]
Thanks @vkk800, we finally got this in. And thanks everyone for the reviews.

**Comment by WarrenWeckesser** [comment]
I added a note about this enhancement to the [release notes](https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.6.0).

**Comment by mdhaber** [comment]
@josef-pkt Should we add tau-a back? It was originally in this PR, but we removed it because tau-b reduces to tau-a in the absence  of ties. But I realize now (e.g. gh-9329) that one could conceivably wish to compute tau-a even in the presence of ties. Is it worth adding it back in?

**Comment by josef-pkt** [comment]
I don't know, I'm currently not looking at association measures and don't know much about different tau versions.

I just came across them (kendall tau and somers d) because they are equivalent to some statistics in the two-sample case.

What I'm trying to figure out is how to compare and extend wilcoxon-mannwhiteney, brunnermunzel and similar rank based statistics.
The statistics I'm looking at all allow for ties because they apply also to discrete, ordinal random variables.
I'm still trying to get an overview and currently have only extensions to brunnermunzel, that's the literature I'm working on right now.

**Comment by dmyersturnbull** [comment]
Which release is this slated for? 1.6.0? I'd like to use it!

**Comment by mdhaber** [comment]
Yes @dmyersturnbull, 1.6.0. 

As a user, I'm curious about your opinion on [this question](https://github.com/scipy/scipy/pull/9361#issuecomment-673864938). Thanks!

**Comment by sappelhoff** [comment]
> As a user, I'm curious about your opinion on this question. Thanks!

for future reader convenience: the question is about "adding back tau-a".

--> I am in favor of that. When I want to use tau-a, it would be convenient to be sure of "using tau-a", without having to check previously whether or not there are ties, and whether tau-b will truly end up being equivalent to tau-a.

As an example for why it's relevant: For the analyses that I am currently running, kendall's tau-a is frequently recommended:

> To judge the ability of a model RDM to explain a brain RDM, we used Kendall’s rank correlation coefficient tau-a (which is the proportion of pairs of values that are consistently ordered in both variables). When comparing models that predict tied ranks (e.g. category model RDMs) to models that make more detailed predictions (e.g. brain RDMs, object-vision model RDMs) Kendall’s tau-a correlation is recommended (Nili et al., 2014), because the Pearson and Spearman correlation coefficients have a tendency to prefer a simplified model that predicts tied ranks for similar dissimilarities over the true model.

- http://www.mrc-cbu.cam.ac.uk/wp-content/uploads/2014/10/Kriegeskorte_Advanced-topics-talk_RSA-Workshop_17-Feb-2015.pdf
- https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3990488/
- https://www.sciencedirect.com/science/article/pii/S0022249616301134

**Comment by sappelhoff** [comment]
https://github.com/scipy/scipy/pull/13855#issuecomment-818832760

> @sappelhoff it can easily be added back. As you see, i was trying to find out whether it should have been added back. Why do you want to calculate it even in the presence of ties? Please reply there, here, or open another issue (and mention me). Thanks!

To be honest, I wasn't even checking my data for the presence or absence of ties, but rather seeing the recommendation to use kendall's tau a for my kind of analysis, so I searched for it ... didn't find it ... and implemented it myself (exactly how you did in https://github.com/scipy/scipy/pull/9361/commits/4d85c35f57cd577a0322becef0d8c7ccb6fdec39 (in that diff you remove it)

The only change being how tau is calculated: `tau = con_minus_dis / np.sqrt(tot - xtie) / np.sqrt(tot - ytie)` --> `tau = con_minus_dis / tot`
I also tested that against an R package (DescTools:KendallTauA) and an implementation from "pyrsa" (https://github.com/rsagroup/pyrsa) (and it worked)

So overall I am not sure whether or not I could simply use kendall's tau b and be fine (would have to check). But I think it'd be good for users to have kendall's tau-a available.

Especially if the added lines / maintenance burden for scipy is so low :slightly_smiling_face: 

I have seen another implementation of this in the wild here: https://github.com/wmvanvliet/mne-rsa/blob/6ad8ea727069fd5bdf0558c4c6324cb87b4ac673/mne_rsa/rsa.py#L16-L19

So I think the need for it is clearly there. Do you maybe have an opinion @wmvanvliet?

If we all agree, I'd be happy to make my first PR to scipy and add kendall's tau a "back"! :wink:

**Comment by wmvanvliet** [comment]
The reason for my implementation was exactly the same as yours @sappelhoff: to do RSA. I've based mine of the version in the MATLAB RSA Toolbox, here: https://github.com/rsagroup/rsatoolbox/blob/develop/%2Brsa/%2Bstat/rankCorr_Kendall_taua.m

It's a big +1 for me to add tau-a to scipy proper. Has the p-value issue been solved?

**Comment by mdhaber** [comment]
What is the p-value issue?

**Pull request merged**
